db/db_sst_test.cc

```diff
@@ -538,6 +538,46 @@ TEST_F(DBSSTTest, DBWithMaxSpaceAllowed) {
   ASSERT_NOK(Flush());
 }
 
+TEST_F(DBSSTTest, CancellingCompactionsWorks) {
+  std::shared_ptr<SstFileManager> sst_file_manager(NewSstFileManager(env_));
+  auto sfm = static_cast<SstFileManagerImpl*>(sst_file_manager.get());
+
+  Options options = CurrentOptions();
+  options.sst_file_manager = sst_file_manager;
+  options.level0_file_num_compaction_trigger = 2;
+  DestroyAndReopen(options);
+
+  int cancelled_compaction=0;
+  rocksdb::SyncPoint::GetInstance()->SetCallBack(
+      "DBImpl::BackgroundCompaction():CancelledCompaction",
+      [&](void* arg) {
+        cancelled_compaction++;
+      });
+  rocksdb::SyncPoint::GetInstance()->EnableProcessing();
+
+  Random rnd(301);
+
+  // Generate a file containing 10 keys.
+  for (int i = 0; i < 10; i++) {
+    ASSERT_OK(Put(Key(i), RandomString(&rnd, 50)));
+  }
+  ASSERT_OK(Flush());
+  uint64_t total_file_size = 0;
+  auto files_in_db = GetAllSSTFiles(&total_file_size);
+  // Set the maximum allowed space usage to the current total size
+  sfm->SetMaxAllowedSpaceUsage(2*total_file_size + 1);
+
+  // Generate another file to trigger compaction.
+  for (int i = 0; i < 10; i++) {
+    ASSERT_OK(Put(Key(i), RandomString(&rnd, 50)));
+  }
+  ASSERT_OK(Flush());
+  dbfull()->TEST_WaitForCompact();
+
+  ASSERT_GT(cancelled_compaction, 0);
+  rocksdb::SyncPoint::GetInstance()->DisableProcessing();
```

Comment:
maybe also good to test that, by the end, the reserved space for compaction is back to zero.

Also have you tested that it can continue after a failure to reserve space? For example, you could call `SetMaxAllowedSpace` in your callback to make the space big enough. (You'd also need to change TEST_WaitForCompact to wait for unscheduled_compactions_ so it doesn't return early like it does now.)

Reply:
Good call, the compaction actually doesn't get rescheduled because the compaction scores are inaccurate.

Reply:
It's probably because we only pick compaction if there's a level with score greater than one. So we need to know score first in order to know we should pick compaction.